Getting Started adjustments, TLC... (#23800)
idaholab/moose: modules/doc/content/getting_started/installation/offline_installation.md

Please be certain, that the machine in which you intend to perform the actual work (which this
document will refer to as the 'target' machine), meets our minimum requirements:

!include sqa/minimum_requirements.md

## Prereqs

!include installation/manual_prereqs.md

## MPI Wrapper

If the target machine is an HPC system, it is likely your system already has an MPI wrapper (and
thus, a GCC or LLVM compiler). Please contact your system administrator, and ask them how to
appropriately make use of that cluster's MPI wrapper. If, by chance you will have to build your own,
you have some choices; [OpenMPI](https://www.open-mpi.org/),
[MVAPICH](https://mvapich.cse.ohio-state.edu/), and [MPICH](https://www.mpich.org/), all work well.
You will want to head on over to one of those product sites, and follow their instructions to build
yourself an MPI wrapper.

!alert note
MVAPICH is regarded as the better wrapper for HPC clusters. However, it is also the most difficult
to build properly for that clusters specific hardware. While you may be able to *build* it easily
enough, you may not have built it *properly* to optimally make use of said cluster's hardware.

## Prepare Directory

> These three libraries can all be obtained from one repository (the moose repository). First, create

Author: Yeah... man I always disliked these instructions. They will always be wrong (the links to those PETSc contribs are always changing).

I almost think we should differ users to Containerized solutions instead. Like Docker. I'll have to think on this. I'll make the changes, but I think in the future this page will be replaced by a containerized solution.

instructions will help you create an environment suitable for MOOSE-based development.

Please be certain, that the machine in which you intend to perform the actual work (which this
document will refer to as the 'target' machine), meets the following requirements:

!include sqa/minimum_requirements.md

# Prerequisites

#### Air-Gapped HPC Clusters

If the target machine is an HPC cluster, it is likely your cluster already has an MPI wrapper (and
thus a GCC or LLVM compiler). Please contact your system administrator and ask them how to
appropriately make use of your cluster's MPI wrapper and compiler. Please note, Intel compilers are
not supported.

#### Personal Workstation

If the target machine is your personal workstation, that machine must sufficiently achieve the
following:

!include installation/manual_prereqs.md

In addition, the target machine will need an MPI wrapper. We recommend one of the following:

- [MPICH](https://www.mpich.org/)

- [OpenMPI](https://www.open-mpi.org/)

!alert note title=Personal Air-Gapped Workstation
Procuring the above on a personal air-gapped workstation will need to be your responsibility. We

> cannot instruct a means for bypassing such security.

Reviewer: not sure what you mean here.
bypass the security of air-gapping? why are we "bypassing" ?

Author: I don't want to instruct people how to create an SSH reverse tunnel. Or how to use/bring in a thumb stick into an obviously required air-gapped secure network...

They will need to figure out how to install these prerequisites on their own. We can't help them `apt-get install`, or `xcode-select`, or use App Store, etc, behind a non-internet accessing computer. This statement basically says: Contact your help desk.

Author: I can reword it a bit? Maybe something less security-like?

```bash
!alert note title=Personal Air-Gapped Workstation
Procuring the above on a machine with no network access is beyond the scope of this document. Please work with your system administrator on satisfying the above prerequisites. 
```

Reviewer: That works

via the same tool used to obtain a developers environment in the Prerequisites section.

!alert note title=Personal Air-Gapped Workstation

> Procuring the above on a machine with no network access is beyond the scope of this document. Please

Reviewer: ```suggestion
Procuring the above on a workstation with no network access is beyond the scope of this document. Please
```

HPC cluster is within right?

Author: I am not sure I understand the question.

HPC is listed on another bullet point. I tried to separate them as best I could:

#### Air-Gapped HPC Clusters

and

#### Personal Workstation

The idea was not to make these too large, or use ## to include them in the TOC (top right table of contents). But make them large enough to separate them as sections.

Author: I thought about indenting them (like we here for each section: https://mooseframework.inl.gov/help/troubleshooting.html) But we quickly run out of space :(

Sometimes I wish we would do away with the left margin!!!

Due to the nature of an offline install, it is possible some tests will fail (missing common system
binaries like `rsync`, or other network related tools not normally present on air-gapped machines).

You will also see SKIPPED tests. This is normal as some tests are specific to Macintosh, some for
Linux. Or some other constraint your machine does not satisfy.

Most tests should pass. If they don't, or you see `MAX FAILURES`, thats a problem! And it needs to
be addressed before continuing. Please supply a report of the actual failure (scroll up a ways). For

> example the following explains little (created with `./run_tests -i always_bad`):

Reviewer: ```suggestion
example the following snippet does not give the full picture (created with `./run_tests -i always_bad`):
```

0 passed, 0 skipped, 0 pending, 2 FAILED
```

> Instead, you need to scroll up and report the actual error:

Reviewer: thanks for including this!

Author: Made this into a template. 👍

make -j 6
```

Again, any errors incurred during this step, is going to be beyond the scope of this document.
Please submit a detailed log of the error, to the
[MOOSE Discussion forum](https://github.com/idaholab/moose/discussions).

## Your Application

With all the libraries built, you are now able to build your application. To recap, there were
several environment variables we previously set, which you will need to always set, when you perform
any MOOSE-based development. Here are those environment variables again (setting them multiple times
is harmless):

```bash
export PATH=/some/path/to/MPI/bin:/some/path/to/GCC/bin:$PATH <--EXAMPLE, CHANGE ME
export CC=mpicc CXX=mpicxx FC=mpif90 F90=mpif90 F77=mpif77
export MOOSE_DIR=$HOME/offline/moose
```

You will need to perform the above, each and every time you log into the target machine and wish to
perform development on your application. To that end, creating a special profile you can source in

> one command is recommended.

Author: I always seem to miss a change  request... Yeah I will add a `!alert tip` here

> one command is recommended.

Reviewer: I was thinking a snippet like:
```
export CC=....
```
and how to export that to a bash script people can run

With MOOSE built and testing successfully, you should now be able to build your application. The
only prerequisites before doing so, is to configure your environment to make use of your MPI
wrapper, and to define where your built copy of MOOSE resides:

```bash
export CC=mpicc CXX=mpicxx FC=mpif90 F90=mpif90 F77=mpif77
export MOOSE_DIR=$HOME/offline/moose
```

You will need to export the above variables each time you open a new terminal window on your HPC
cluster or personal workstation, before you can perform any application development.

For this reason you may wish to add the above export commands to your shell startup profile so that
these variables are always set. Achieving this is different for each shell type (and there are a
lot). You'll want to read up on how to do this for Macintosh (usually ZSH: `~/.zshrc`), or Linux

> (usually BASH: `~/.bashrc`) respectfully.

Reviewer: ```suggestion
(usually BASH: `~/.bashrc`) respectively.
```
